src/main/kotlin/cacophony/pipeline/CacophonyPipeline.kt

```diff
@@ -235,4 +237,40 @@ class CacophonyPipeline(
         asm.forEach { (function, asm) -> println("$function generates asm:\n$asm") }
         return asm
     }
+
+    fun generateAsm(input: Input, dest: Path) {
+        val defs = generateAsm(input)
+        Files.newBufferedWriter(dest, StandardOpenOption.CREATE).use { writer ->
+            defs.values.forEach {
+                writer.write(it)
+                writer.newLine()
+            }
+        }
+    }
+
+    fun compile(src: Path, dest: Path) {
+        val nasm = ProcessBuilder("nasm", "-f", "elf64", "-o", dest.toString(), src.toString()).inheritIO().start()
+        nasm.waitFor().takeIf { it != 0 }?.let { status ->
+            logger?.logFailedAssembling(status)
+            throw CompileException("Unable to assemble generated code")
+        } ?: logger?.logSuccessfulAssembling(dest)
+    }
+
+    fun link(src: Path, dest: Path) {
+        val gcc = ProcessBuilder("gcc", "-no-pie", "-o", dest.toString(), src.toString()).inheritIO().start()
+        gcc.waitFor().takeIf { it != 0 }?.let { status ->
+            logger?.logFailedLinking(status)
+            throw CompileException("Unable to link compiled code")
+        } ?: logger?.logSuccessfulLinking(dest)
+    }
+
+    fun compile(input: Input, src: Path) {
+        val asmFile = withExtension(src.fileName, ".asm")
+        val objFile = withExtension(src.fileName, ".o")
+        val binFile = withExtension(src.fileName, "")
```

Comment:
This fails on files with no extensions. Personally I see no reason to strip the extension by default (unless we force the input extension to always be the same), imo it is better to just append new extension to the end of old file name.